File: leetcode/leetcode_2397.cpp
/**
 * @file leetcode_2397.cpp
 * @author Cattle_Horse (CattleHorse@duck.com)
 * @brief https://leetcode.cn/problems/maximum-rows-covered-by-columns
 * @version 0.1
 * @date 2024-05-31
 *
 * @copyright Copyright (c) 2024
 *
 */

#include <algorithm>
#include <vector>

// Gosper's Hack : https://zhuanlan.zhihu.com/p/360512296
// https://programmingforinsomniacs.blogspot.com/2018/03/gospers-hack-explained.html
/*
int m = 1<<n
for (auto i{(1u << k) - 1}; i < m; i = next_k_subset(i)) {
    // Any i contains k ones
    ...
}
*/
#define next_k_subset(i) (((i ^ (i + (i & -i))) >> (std::countr_zero(i & -i) + 2)) | (i + (i & -i)))

class Solution {
public:
    int maximumRows(std::vector<std::vector<int>>& matrix, int numSelect) {
        int m = matrix.size(), n = matrix.front().size();
        std::vector<int> mask(m);
        for (int i{0}; i < m; ++i) {
            for (int j{0}; j < n; ++j) {

                mask[i] |= matrix[i][j] << j;
            }
        }

        int ans = 0;
        for (auto i = (1u << numSelect) - 1, s = 1u << n; i < s; i = next_k_subset(i)) {
            ans = std::max(ans, static_cast<int>(std::count_if(mask.begin(), mask.end(), [&](int v) {
                               return (v & i) == v;
                           })));
        }
        return ans;
    }
};

File: leetcode/leetcode_2397.py
"""
https://leetcode.cn/problems/maximum-rows-covered-by-columns/description/
2397. 被列覆盖的最多行数
"""

from typing import List
from typing import Generator


def next_k_subset(s: int) -> int:
    """下一个k子集(用于k_subset方法)
    https://leetcode.cn/problems/maximum-rows-covered-by-columns
    Gosper's Hack : https://zhuanlan.zhihu.com/p/360512296
    https://programmingforinsomniacs.blogspot.com/2018/03/gospers-hack-explained.html
    """
    lb = s & -s
    r = s + lb
    return ((r ^ s) // lb >> 2) | r


def k_subset(S: int, k: int) -> Generator[int, None, None]:
    """生成子集(每一个集合都有k个元素)"""
    s = (1 << k) - 1
    while s <= S:
        yield s
        s = next_k_subset(s)


class Solution:
    def maximumRows(self, matrix: List[List[int]], numSelect: int) -> int:
        m, n = len(matrix), len(matrix[0])
        mask = [0] * m
        for i, row in enumerate(matrix):
            for j, val in enumerate(row):
                mask[i] |= val << j
        count_if = lambda s: sum(1 for v in mask if v & s == v)
        return max(map(count_if, k_subset(1 << n, numSelect)))
